src/LanguageServer/Protocol/Features/Options/SolutionAnalyzerConfigOptionsUpdater.cs

```diff
@@ -0,0 +1,273 @@
+
+namespace Microsoft.CodeAnalysis.Options;
+
+/// <summary>
+/// Keeps <see cref="SolutionState.FallbackAnalyzerOptions"/> up-to-date with global option values maintained by <see cref="IGlobalOptionService"/>.
+/// </summary>
+[Export]
+[ExportEventListener(WellKnownEventListeners.Workspace, WorkspaceKind.Host, WorkspaceKind.Interactive, WorkspaceKind.SemanticSearch), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class SolutionAnalyzerConfigOptionsUpdater(
+    EditorConfigOptionsEnumerator optionsEnumerator,
+    IGlobalOptionService globalOptions) : IEventListener<object>, IEventListenerStoppable
+{
+    /// <summary>
+    /// We need to update the current solution snapshot based on two sources of events:
+    /// 1) global options service (global option changed)
+    /// 2) the workspace (project added)
+    /// 
+    /// To keep the set of analyzer fallback options on the solution snapshot consistent and not miss any option 
+    /// we serialize the updates using a lock.
+    /// </summary>
+    private sealed class WorkspaceUpdater(Workspace workspace, EditorConfigOptionsEnumerator optionsEnumerator, IGlobalOptionService globalOptions)
+    {
+        // Set of languages (keys of the dictionary) in the current snapshot that have their fallback options initialized from global options.
+        // The values are ignored.
+        private ImmutableDictionary<string, int> _initializedLanguages = ImmutableDictionary<string, int>.Empty;
+
+        // Ensures we only process one update at a time:
+        private readonly object _updateLock = new();
+
+        public void WorkspaceChanged(object? sender, WorkspaceChangeEventArgs args)
+        {
+            try
+            {
+                switch (args.Kind)
+                {
+                    case WorkspaceChangeKind.SolutionAdded:
+                    case WorkspaceChangeKind.SolutionChanged:
+                    case WorkspaceChangeKind.SolutionCleared:
+                    case WorkspaceChangeKind.SolutionReloaded:
+                    case WorkspaceChangeKind.SolutionRemoved:
+                    case WorkspaceChangeKind.ProjectAdded:
+                    case WorkspaceChangeKind.ProjectChanged:
+                    case WorkspaceChangeKind.ProjectReloaded:
+                    case WorkspaceChangeKind.ProjectRemoved:
+                        break;
+
+                    default:
+                        return;
+                }
+
+                if (args.OldSolution.SolutionState.ProjectCountByLanguage.KeysEqual(args.NewSolution.SolutionState.ProjectCountByLanguage))
+                {
+                    return;
+                }
+
+                UpdateSolution(
+                    transformation: solution => InitializeLanguages(solution, _initializedLanguages),
+                    onAfterUpdate: solution => _initializedLanguages = solution.SolutionState.ProjectCountByLanguage);
```

Comment:
not sure how i feel about onAfterUpdate being used in this fashion.  not totally opposed to it.  alternatively, consider if UpdateSOlution returned the new solution and you just assigned _initializedLanguages based on that.